pymc_marketing/mmm/base.py

```diff
@@ -78,132 +67,241 @@ def methods(self) -> List[Any]:
 
-    def get_target_transformer(self) -> Pipeline:
-        try:
-            return self.target_transformer  # type: ignore
-        except AttributeError:
-            identity_transformer = FunctionTransformer()
-            return Pipeline(steps=[("scaler", identity_transformer)])
+        This property scans the methods of the object and returns those marked for preprocessing.
+        The methods are marked by having a _tags dictionary attribute, with either 'preprocessing_X' or 'preprocessing_y' set to True.
+        The 'preprocessing_X' tag indicates a method used for preprocessing features, and 'preprocessing_y' indicates a method used for preprocessing the target variable.
 
-    def validate(self, data: pd.DataFrame):
-        for method in self.validation_methods:
-            method(self, data)
+        Returns
+        -------
+        tuple of list of Callable[["BaseMMM", pd.DataFrame], pd.DataFrame]
+            A tuple where the first element is a list of methods for 'X' preprocessing, and the second element is a list of methods for 'y' preprocessing.
 
-    def preprocess(self, data: pd.DataFrame) -> pd.DataFrame:
-        for method in self.preprocessing_methods:
-            data = method(self, data)
-        return data
+        Example
+        -------
+        >>> self.preprocessing_methods
+        """
+        return (
+            [
+                method
+                for method in self.methods
+                if getattr(method, "_tags", {}).get("preprocessing_X", False)
+            ],
+            [
+                method
+                for method in self.methods
+                if getattr(method, "_tags", {}).get("preprocessing_y", False)
+            ],
+        )
+
+    def preprocess(
+        self, target: str, data: Union[pd.DataFrame, pd.Series]
+    ) -> Union[pd.DataFrame, pd.Series]:
+        """
+        Preprocess the provided data according to the specified target.
+
+        This method applies preprocessing methods to the data ('X' or 'y'), which are specified in the preprocessing_methods property of this object.
+        It iteratively applies each method in the appropriate list (either for 'X' or 'y') to the data.
+
+        Parameters
+        ----------
+        target : str
+            Indicates whether the data represents features ('X') or the target variable ('y').
 
-    @abstractmethod
-    def build_model(self, *args, **kwargs) -> None:
-        raise NotImplementedError()
+        data : pd.DataFrame
+            The data to be preprocessed.
 
-    def get_prior_predictive_data(self, *args, **kwargs) -> az.InferenceData:
+        Returns
+        -------
+        pd.DataFrame
+            The preprocessed data.
+
+        Raises
+        ------
+        ValueError
+            If the target is neither 'X' nor 'y'.
+
+        Example
+        -------
+        >>> data = pd.DataFrame({"x1": [1, 2, 3], "y": [4, 5, 6]})
+        >>> self.preprocess('X', data)
```

Comment:
Can we use " for strings? Should be `self.preprocess ("X", data)`

Reply:
sure, I'll try to pay attention to that from now on, I'm 'energy efficient' so when I need to press only one button I choose ' over " intuitively :)

Reply:
I've replaced all ' for " (where it was possible)